### ISSUE TITLE
parser: error parsing functions

### ISSUE THREAD

**Issue by Khasms** [typescript-eslint/typescript-eslint#3965]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```TS
const { parse } = require('@typescript-eslint/parser');
parse('(a) => a')
```

**Expected Result**

The parser does not throw an error.

**Actual Result**

The parser throws an error.
```sh
E:\test\node_modules\.pnpm\@typescript-eslint+scope-manager@4.33.0\node_modules\@typescript-eslint\scope-manager\dist\scope\FunctionScope.js:28
        const bodyStart = (_b = (_a = this.block.body) === null || _a === void 0 ? void 0 : _a.range[0]) !== null && _b !== void 0 ? _b : -1;
                                                                                                    ^

TypeError: Cannot read property '0' of undefined
    at FunctionScope.isValidResolution (E:\test\node_modules\.pnpm\@typescript-eslint+scope-manager@4.33.0\node_modules\@typescript-eslint\scope-manager\dist\scope\FunctionScope.js:28:101)       
    at resolve (E:\test\node_modules\.pnpm\@typescript-eslint+scope-manager@4.33.0\node_modules\@typescript-eslint\scope-manager\dist\scope\ScopeBase.js:182:27)
    at E:\test\node_modules\.pnpm\@typescript-eslint+scope-manager@4.33.0\node_modules\@typescript-eslint\scope-manager\dist\scope\ScopeBase.js:195:18
    at FunctionScope.close (E:\test\node_modules\.pnpm\@typescript-eslint+scope-manager@4.33.0\node_modules\@typescript-eslint\scope-manager\dist\scope\ScopeBase.js:286:13)
    at Referencer.close (E:\test\node_modules\.pnpm\@typescript-eslint+scope-manager@4.33.0\node_modules\@typescript-eslint\scope-manager\dist\referencer\Referencer.js:51:66)
    at Referencer.visitFunction (E:\test\node_modules\.pnpm\@typescript-eslint+scope-manager@4.33.0\node_modules\@typescript-eslint\scope-manager\dist\referencer\Referencer.js:191:14)
    at Referencer.ArrowFunctionExpression (E:\test\node_modules\.pnpm\@typescript-eslint+scope-manager@4.33.0\node_modules\@typescript-eslint\scope-manager\dist\referencer\Referencer.js:213:14)  
    at Referencer.visit (E:\test\node_modules\.pnpm\@typescript-eslint+scope-manager@4.33.0\node_modules\@typescript-eslint\scope-manager\dist\referencer\VisitorBase.js:72:21)
    at Referencer.visitChildren (E:\test\node_modules\.pnpm\@typescript-eslint+scope-manager@4.33.0\node_modules\@typescript-eslint\scope-manager\dist\referencer\VisitorBase.js:59:22)
    at Referencer.visit (E:\test\node_modules\.pnpm\@typescript-eslint+scope-manager@4.33.0\node_modules\@typescript-eslint\scope-manager\dist\referencer\VisitorBase.js:77:14)
```
**Additional Info**

This only errors if it is parsing a function in which at least one parameter is used. For example, parsing `(a) => z` will not error, but `(a) => a` will. Errors for both arrow functions and regular functions. I've tested on node 14 and 16 and every `parser` version back to the latest v3 release. I am getting this error using `parser` but the error seems to be originating from `scope-manager` so sorry if I labeled it wrong.

**Versions**

| package                     | version |
| --------------------------- | ------- |
| `@typescript-eslint/parser` | `4.33.0` |
| `TypeScript`                | `4.4.3` |
| `ESLint`                    | `7.32.0` |
| `node`                      | `16.6.0` |

**Comment by bradzacher** [comment]
`@typescript-eslint/parser` isn't really made for programatic public consumption - it's really supposed to be used by eslint directly.

Instead you should use `@typescript-eslint/typescript-estree` if you want to programatically use a parser.

If for some reason you must use `/parser` - then you must turn on the `range` option.
